### PR TITLE
Fix object_id usage as thread local key

### PR DESCRIPTION
**Pull request by marcotc**
I stumbled upon this flaky error: https://app.circleci.com/pipelines/github/DataDog/dd-trace-rb/6331/workflows/02b56e62-53d7-4d50-ac12-f3e84194bf8f/jobs/234310/parallel-runs/1
```
Failures:

  1) Datadog::Tracing::FiberLocalContext#initialize create one fiber-local variable
     Failure/Error: expect { fiber_local_context }.to change { Thread.current.keys.size }.by(1)
       expected `Thread.current.keys.size` to have changed by 1, but was changed by 0
     # ./spec/datadog/tracing/context_provider_spec.rb:100:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:216:in `block (2 levels) in <top (required)>'
     # ./spec/spec_helper.rb:108:in `block (2 levels) in <top (required)>'
     # /usr/local/bundle/gems/webmock-3.13.0/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'
```

It happens because we use `object_id` as a thread local key and `object_id` can be reused after garbage collection, causing collision between a recently garbage collected instance of `FiberLocalContext` and a new one.

This PR addresses this issue in two places where we use `object_id` as a thread local key by creating a thread-safe counter for each of the object instances.

**If there's an easier/cheaper way to get a unique identifier please let me know!**

### CODE REVIEW

**Comment by ivoanjo** [comment]
> If there's an easier/cheaper way to get a unique identifier please let me know!

I tried spending a bit of time thinking about this. There's quite a lot of abstraction layers here, but here's my current understanding:

* A given `Datadog::Tracing::Tracer` gets initialized with a `context_provider`, which defaults to `Datadog::Tracing::DefaultContextProvider`. So there's a 1:1 relationship here.
* Then the `DefaultContextProvider` keeps a `@context` which is a `FiberLocalContext`, again a 1:1 relationship.
* Each `FiberLocalContext` wants to have a unique key (`@key`), which it was (incorrectly) deriving from its `object_id`.
* Inside `FiberLocalContext`, we use this unique key as a key to `Thread.current`, and use it to store a `Datadog::Tracing::Context` instance that is unique for every Thread/Fiber.

Flattening a bit down the layers, it appears to me that the intention here is to effectively have a unique key for each instance of `Datadog::Tracing::Tracer`.

But since this is a unique key per tracer instance, and we don't expect many tracer instances to exist in a given application -- because if we do, we need to start thinking about cleaning up leftover contexts because otherwise they will lead to a memory leak -- I don't think we need a cheaper (performance-wise) way of doing this, since the common situation is that we'll do this one or at most a handful of times in practice.

In terms of easier way, I don't think there is: https://bugs.ruby-lang.org/issues/12607 is actually what we'd want (atomic integer) but hasn't been accepted yet.

**Comment by ivoanjo** [comment]
Also, this is not in the part of the code you changed, but is this a potential bug?

https://github.com/DataDog/dd-trace-rb/blob/06561981b77c1db5c5a70af85d3cc76e2a054322/lib/datadog/tracing/context_provider.rb#L23-L34

In particular, when key is **another thread**, it seems weird to then say "whatever trace that other thread was running, I'm also running it now".

Should we only run the `after_fork!` block when getting the context of the current thread? (e.g. `key == nil || key == Thread.current`?)

File: lib/datadog/tracing/context_provider.rb [hunk 1]
```diff
@@ -61,6 +62,18 @@ def local=(ctx)
...
+      # We previously used {FiberLocalContext#object_id} to ensure uniqueness
+      # but the VM is allowed to reuse `object_id`, allow for the possibility that
+      # a new FiberLocalContext would be able to read an old FiberLocalContext's
+      # value.
+      @mutex = Mutex.new
+      @unique_instance_id = Datadog::Core::Utils::Sequence.new
+
+      def self.next_instance_id
+        @mutex.synchronize { @unique_instance_id.next }
+      end
```
Comment (ivoanjo) [inline lib/datadog/tracing/context_provider.rb:76]: Minor: I usually favor putting these into constants, to make it even more clear they are not instance-scoped, but the effect is exactly the same so it's really just for readibility.

```suggestion
      UNIQUE_INSTANCE_MUTEX = Mutex.new
      UNIQUE_INSTANCE_GENERATOR = Datadog::Core::Utils::Sequence.new

      def self.next_instance_id
        UNIQUE_INSTANCE_MUTEX.synchronize { UNIQUE_INSTANCE_GENERATOR.next }
      end
```

(Potentially add a `private_constant :UNIQUE_INSTANCE_MUTEX` + generator as well maybe?)

**Review by ivoanjo — APPROVED** [review]
Seems reasonable to me!

**Comment by marcotc** [comment]
> Should we only run the after_fork! block when getting the context of the current thread? (e.g. key == nil || key == Thread.current?)

This seems reasonable. I'll add a comment for this.

File: lib/datadog/tracing/context_provider.rb (hunk 1 above)
Comment (marcotc) [reply lib/datadog/tracing/context_provider.rb]: I don't really mind either way, but I hate to disappoint a reviewer :)

**Pull request merged**